pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/segmentrefresh/SegmentRefreshTaskExecutor.java

```diff
@@ -0,0 +1,200 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.plugin.minion.tasks.segmentrefresh;
+
+import java.io.File;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import org.apache.pinot.common.metadata.segment.SegmentZKMetadataCustomMapModifier;
+import org.apache.pinot.core.common.MinionConstants;
+import org.apache.pinot.core.minion.PinotTaskConfig;
+import org.apache.pinot.plugin.minion.tasks.BaseSingleSegmentConversionExecutor;
+import org.apache.pinot.plugin.minion.tasks.SegmentConversionResult;
+import org.apache.pinot.segment.local.indexsegment.immutable.ImmutableSegmentLoader;
+import org.apache.pinot.segment.local.segment.creator.impl.SegmentIndexCreationDriverImpl;
+import org.apache.pinot.segment.local.segment.index.loader.IndexLoadingConfig;
+import org.apache.pinot.segment.local.segment.readers.PinotSegmentRecordReader;
+import org.apache.pinot.segment.spi.ColumnMetadata;
+import org.apache.pinot.segment.spi.creator.SegmentGeneratorConfig;
+import org.apache.pinot.segment.spi.index.metadata.SegmentMetadataImpl;
+import org.apache.pinot.segment.spi.loader.SegmentDirectoryLoaderContext;
+import org.apache.pinot.segment.spi.loader.SegmentDirectoryLoaderRegistry;
+import org.apache.pinot.segment.spi.store.SegmentDirectory;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class SegmentRefreshTaskExecutor extends BaseSingleSegmentConversionExecutor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentRefreshTaskGenerator.class);
+
+  private long _taskStartTime;
+
+  /**
+   * The code here currently covers segmentRefresh for the following cases:
+   * 1. Process newly added columns.
+   * 2. Addition/removal of indexes.
+   * 3. Compatible datatype change for existing columns
+   */
+  @Override
+  protected SegmentConversionResult convert(PinotTaskConfig pinotTaskConfig, File indexDir, File workingDir)
+      throws Exception {
+    _eventObserver.notifyProgress(pinotTaskConfig, "Refreshing segment: " + indexDir);
+
+    // We set _taskStartTime before fetching the tableConfig. Task Generation relies on tableConfig/Schema updates
+    // happening after the last processed time. So we explicity use the timestamp before fetching tableConfig as the
+    // processedTime.
+    _taskStartTime = System.currentTimeMillis();
```

Comment:
There are pros and cons to either approach:

1. With the current way - we pay the computation cost on the Minions (and the controllers/helix). 
2. If we use `getSegmentsMetadata` we will pay computation cost on the Pinot Servers. 

I resolved to (1) to avoid overloading the servers. Do you see any compelling reason for going with approach (2) and issuing a call to Pinot Servers?

Reply:
For 2), the cost should be minimal since the API will be called once every scheduled trigger. However we are shipping the metadata payload to the controller which can add to the controller heap (if you have 1000's of segments). This is one downside. The controller/task generator uses the metadata to check if a segment has diverged from table config. 

The reason I suggested that is we will know exactly whether subtasks need to be triggered or not, instead of triggering subtasks for every table update even if its unrelated (if I understood the trigger logic).

Reply:
I thought about this. The compute/memory on minion is cheaper and I'm inclined to doing it this way. Moreover, doing it at minion gives us the flexibility of also support compression type changes on segments (where we have to initialize the forward index reader to get the compression info)

Reply:
My concerns are, we are unnecessarily doing the following
1. Creating a large number of subtasks when refresh is not required
2. BaseSingleSegmentConversionExecutor modifies Zk metadata and even uploads the segment, which is not necessary when refresh for that segment is not required. Segment requires no changes. 

@vrajat is working on the server side API to detect diverged segments. This will reduce the payload by not shipping the segment metadata/but only sending the segment names that require refresh. The call to the server will only happen during task scheduled trigger (order of minutes/hours) which in my opinion should not put load on the server. Until then we can use getSegmentMetadata call on the server?
 
https://github.com/apache/pinot/issues/14450

Reply:
I see recently we added an API on server / controller to see if reload is required or not: #12117

Reply:
Apologies for the pass through comment. I've updated the description with the reasoning for not using server APIs to detect this.  Please take a look at the "Notes on Implementation" section. @swaminathanmanish  I've also pinged on slack to quickly align on this.

Reply:
Synced with @swaminathanmanish offline. We'll go with these changes now to avoid the controller heap blowup by transporting all segment metadata. We can use the API in https://github.com/apache/pinot/issues/14450 once it is merged.